src/helper/remote_config/client.cpp

```diff
@@ -0,0 +1,218 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+#include "client.hpp"
+#include "product.hpp"
+#include "protocol/tuf/parser.hpp"
+#include "protocol/tuf/serializer.hpp"
+#include <algorithm>
+#include <map>
+#include <regex>
+
+namespace dds::remote_config {
+
+protocol::remote_config_result config_path_from_path(
+    const std::string &path, config_path &cp)
+{
+    std::regex regex("^(datadog/\\d+|employee)/([^/]+)/([^/]+)/[^/]+$");
+
+    std::smatch base_match;
+    if (!std::regex_match(path, base_match, regex) || base_match.size() < 4) {
+        return protocol::remote_config_result::error;
+    }
+
+    std::string product(base_match[2].str());
+    std::string id(base_match[3].str());
+    cp.set_product(product);
+    cp.set_id(id);
+
+    return protocol::remote_config_result::success;
+}
+
+protocol::get_configs_request client::generate_request()
+{
+    std::vector<remote_config::protocol::config_state> config_states;
+    std::vector<protocol::cached_target_files> files;
+
+    for (std::pair<std::string, product> pair : this->_products) {
+        // State
+        auto configs_on_product = pair.second.get_configs();
+        for (config _c : configs_on_product) {
+            std::string c_id(_c.get_id());
+            std::string c_product(_c.get_product());
+            remote_config::protocol::config_state cs(
```

Comment:
This is a similar situation as before, a copy of a string which is then immediately discarded, I recommend making `config_state` take rvalue references and std::move internally..

Reply:
Done